File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBoundedValidBST(TreeNode root, Integer left, Integer right) {
        if (root == null) return true;

        if (!(root.val > left && root.val < right)) {
            return false;
        }

        return isBoundedValidBST(root.left, left, root.val) && 
               isBoundedValidBST(root.right, root.val, right);

    }
    public boolean isValidBST(TreeNode root) {
        return isBoundedValidBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    /*pair<bool, int> helper(TreeNode* root) {
        if (!root) return {true, 0};
        auto [lb, lh] = helper(root->left);
        auto [rb, rh] = helper(root->right);
        int height = 1 + max(lh, rh);
        bool balanced = lb && rb && (abs(lh - rh) <= 1);
        return {balanced, height};
    }
    bool isBalanced(TreeNode* root) {
        return helper(root).first;
    }*/
    int helper(TreeNode* root) {
        if (!root) return 0;
        int lh = helper(root->left);
        if (lh < 0) return -1;
        int rh = helper(root->right);
        if (rh < 0) return -1;
        if (abs(lh - rh) > 1) return -1;
        int height = 1 + max(lh, rh);
        return height;
    }
    bool isBalanced(TreeNode* root) {
        return (helper(root) >= 0);
    }
};
